### ISSUE TITLE
group membership (discussion)

### ISSUE THREAD

**Issue by arubdesu** [facebook/osquery#1762]
There's an existing master task on #619 under Shared (cross platform) tables for group membership. To expand on scenarios/actionable criteria worth collecting, just sticking to the native local directory service, it would be vital to know who is in the admin group, and further to know(/potentially use as a foreign key?) nested groups - e.g., `staff` and or `netaccounts` in the `_lpadmin` group, to allow 'standard' (non-admin) users to print on Darwin. There is also a system ssh group, com.apple.access_ssh (whose GID has varied through different OS releases) that it is a common pattern to nest the admin group in.

**Comment by theopolis** [comment]
If you want to know who's in a group:
```sql
select * from groups join user_groups using (gid) join users using (uid) where groupname = 'staff';
```

That works for a single group name, I'll have to work on optimizing when using SQL's nice `IN` or `OR` (for some reason `OR`s make constraints un-indexable).

**Comment by arubdesu** [comment]
I think I'm asking for nested_gid schema to be added to the user_groups table, and then `join`s can resolve the full actual memberships?

**Comment by theopolis** [comment]
I'm still stuck on this. With `groups` (`gid`) `user_groups` (`gid`, `uid`) and `users` (`uid`), we should be able to go both ways:
```
osquery> select uid, gid, groupname from groups join user_groups using (gid) where groupname = 'com.apple.access_ssh';
+-----------+-----+----------------------+
| uid       | gid | groupname            |
+-----------+-----+----------------------+
| 499       | 399 | com.apple.access_ssh |
| 80        | 399 | com.apple.access_ssh |
| 329559041 | 399 | com.apple.access_ssh |
| 0         | 399 | com.apple.access_ssh |
+-----------+-----+----------------------+
```

I'm not the swiftest tool in the shed. If you don't mind explaining the join we cannot perform with those three tables I can take it from there!

**Comment by arubdesu** [comment]
Here's what I'm missing:
```
$ dscl . read /Groups/_lpadmin
AppleMetaNodeLocation: /Local/Default
GeneratedUID: ABCDEFAB-CDEF-ABCD-EFAB-CDEF00000062
GroupMembership: localabanks
NestedGroups: ABCDEFAB-CDEF-ABCD-EFAB-CDEF00000050 ABCDEFAB-CDEF-ABCD-EFAB-CDEF0000003E ABCDEFAB-CDEF-ABCD-EFAB-CDEF00000014
Password: *
PrimaryGroupID: 98
RealName:
 Print Administrators
RecordName:
 _lpadmin
 lpadmin
 BUILTIN\Print Operators
RecordType: dsRecTypeStandard:Groups
SMBSID: S-1-5-32-550
```
The NestedGroups above are the only way I can tell that two groups, in this case `netaccounts` and `staff`, can inherit rights assigned by the authorization framework to modify printers, a system-wide privilege. I can't go backwards from the group itself to know that.
```
$ dscl . read /Groups/netaccounts
AppleMetaNodeLocation: /Local/Default
Comment:
 Group membership calculated by system
 Accounts from a remote directory server
GeneratedUID: ABCDEFAB-CDEF-ABCD-EFAB-CDEF0000003E
Password: *
PrimaryGroupID: 62
RealName:
 Network Accounts
RecordName: netaccounts
RecordType: dsRecTypeStandard:Groups
```

**Comment by arubdesu** [comment]
In other words, users know what groups they're in, groups know who their (user) members are, but groups don't currently have schema for nested groups. There's no join I can perform (I think) at present to get at that relationship.

**Comment by theopolis** [comment]
Ah! Thanks @arubdesu, I understand now. For some reason I wasn't connecting with the nested groups concept or where/how those relationships were stored-- I was stuck in what was possible/reported from POSIX.

**Comment by arubdesu** [comment]
This is now clearly in my sights to make a POC with the python extension module

**Comment by arubdesu** [comment]
OK, POC created: https://gist.github.com/arubdesu/29cd6dc7eb915c91c1586fb17e99b74f This mockup has the benefit of not shelling out at all, so the frameworks may be a easier thing to implement. Not sure if this wants to be munged with the current `groups` table implementation when on darwin or have it supercede the original for macOS.

**Comment by arubdesu** [comment]
(Could I get the question label removed and wishlist applied please?)

**Comment by fmanco** [comment]
@arubdesu are you still interested in this and does this issue still make sense. If yes I'm going to keep open with the `wishlist` label. Otherwise I'm going to close.

**Comment by arubdesu** [comment]
Yes I’m still interested, I’m not sure if other platforms have nesting like Darwin, but I think it’s worth it since it’s a part of access control

**Comment by fmanco** [comment]
Okay. Unfortunately we don't currently have plans to work on it but if this is still interesting I'll leave it open. I also updated the title to make it more explicit, but feel free to suggest a better one.

**Comment by toliver38** [comment]
I'm keen to see if this is still worth considering. I use a similar query type to track users by group but this doesn't truely reflect the the scope in reality given the nested group problem. Specifically I am trying to identify a way to identify if a user has been added to a group that is a sub group of a more privileged parent group such as Domain Admin
